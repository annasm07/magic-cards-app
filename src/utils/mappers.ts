import { Card, CardDTO } from "@/types/Card";

const DEFAULT_CARD_IMAGE =
  "https://propagandamtg.com/cdn/shop/files/mtg_placeholder_2bfc0f4d-50f6-4158-9104-ffb14d2fe7c5_800x.png?v=1736530948";

export const mapCardAPItoVM = (card: CardDTO): Card => {
  return {
    id: card.id,
    name: card.name,
    type: card.type,
    manaCost: card.manaCost,
    rarity: card.rarity,
    set: card.setName,
    illustrator: card.artist,
    text: card.text,
    colors: card.colors,
    imageUrl: card.imageUrl || DEFAULT_CARD_IMAGE,
  };
};

export const mapToCollection = <Item, ItemResult>(
  collection: Item[],
  mapItemFn: (item: Item, index?: number) => ItemResult
): ItemResult[] => (Array.isArray(collection) ? collection.map(mapItemFn) : []);
