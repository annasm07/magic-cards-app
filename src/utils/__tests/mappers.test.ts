import { mockedCardsDTO } from "@/__mocks__/cards.mock";
import { mapCardAPItoVM, mapToCollection } from "../mappers";

describe("Mappers", () => {
  describe("mapToCollection", () => {
    it("should return a card model with the correct properties", () => {
      const card = mockedCardsDTO[0];

      const cardVm = mapCardAPItoVM(card);

      expect(cardVm).toMatchSnapshot();
    });

    it("should return a card model with the default values", () => {
      const card = mockedCardsDTO[0];

      card.imageUrl = undefined;

      const cardVm = mapCardAPItoVM(card);

      expect(cardVm.imageUrl).toEqual(
        "https://propagandamtg.com/cdn/shop/files/mtg_placeholder_2bfc0f4d-50f6-4158-9104-ffb14d2fe7c5_800x.png?v=1736530948"
      );
    });
  });
  describe("mapToCollection", () => {
    // @ts-ignore - Proposal tests.
    const invalidCases = [null, undefined, []];
    it.each(invalidCases)(
      "should return an empty string when passes %p",
      (testValue) => {
        const irrelevantMapper = (value: string) => value;
        expect(mapToCollection(testValue, irrelevantMapper)).toEqual([]);
      }
    );

    it("should map correctly", () => {
      const irrelevantMapper = (value: number) => value + 1;
      expect(mapToCollection([1, 2, 3], irrelevantMapper)).toEqual([2, 3, 4]);
    });
  });
});
