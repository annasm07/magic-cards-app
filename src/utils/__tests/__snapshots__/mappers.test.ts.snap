// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mappers mapToCollection should return a card model with the correct properties 1`] = `
{
  "colors": [
    "W",
  ],
  "id": "5f8287b1-5bb6-5f4c-ad17-316a40d5bb0c",
  "illustrator": "Pete Venters",
  "imageUrl": "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=130550&type=card",
  "manaCost": "{5}{W}{W}",
  "name": "Ancestor's Chosen",
  "rarity": "Uncommon",
  "set": "Tenth Edition",
  "text": "First strike (This creature deals combat damage before creatures without first strike.)
When Ancestor's Chosen enters the battlefield, you gain 1 life for each card in your graveyard.",
  "type": "Creature â€” Human Cleric",
}
`;
